text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,443 @@
+
+It is argued here that an iterator using the name `Range` should produce a sequence of values that lie within the `min, max` bounds. Therefore, most of what are now *infinite* ranges should instead be *empty* ranges (where `.has_next() == false`). The RFC proposes a delineation of the so-far *infinite* cases into one of two types: those that are now *empty*, and those which remain *infinite*.
+
+A smaller issue proposed to be fixed by this RFC is the naming of the Range bounds `min` and `max`. Especially when an implementation allows for descending Ranges like Pony's does -- e.g. `Range(-10, -20, -1)`, naming the first parameter `min` and the second `max` is clearly mathematically wrong. Therefore if accepted, in the actual implementation these will be renamed to `start` and `end`. However, to not confuse things, they will still be named `min, max` throughout here.
+
+# Motivation
+
+Currently, a Range is considered infinite if either 1) the `step` is `0`, or any of `min`, `max`, or `step` are `NaN`, `+Inf` or `-Inf`, or 2) if no progress can be made from `min` to `max` due to the sign of the `step`.
+
+In this RFC it will be argued that 1) is too broad in that it creates infinite iterators that should be *empty*, and that 2) should be entirely carved out from the set of currently *infinite* iterators all of which should also instead be *empty*. Currently, the following code fragment
+
+```pony
+  let r =  Range[I32](0, 10, -1)
+  for i in Range(0, 10) do
+    try env.out.print(r.next()?.string()) end
+  end
+```
+
+produces
+
+```
+0
+-1
+-2
+-3
+-4
+-5
+-6
+-7
+-8
+-9
+```
+
+All of these but the first value lie outside the range 0..10 that the user requested the Range class to generate. However, not the least based on the word *Range* -- defined by Merriam Webster as ".. 7a: a sequence, series, or scale between limits" -- the user can expect this to produce an iterator that generates all values along the *trajectory* from `min` to `max`, interspaced by `step`. Therefore, no sequence value should lie *outside* `[min, max)`.
```

Comment:
This definition excludes "infinitely" producing the same value. I agree that Range should not include such a use case. We might want to add a suggestion -- not necessarily where I am making this comment -- that if someone needs a infinite production of the same value to use "itertools" `Iter[A].repeat_value(value)`.

Reply:
Will do.